### PR TITLE
Reply to comment (merge comments + answers) feature

### LINKED ISSUE #4094: Planning issue: merging Answers into Comments
There is a long list of issues with Answers and Comments, not only because they're easy to confuse (#884) but also because having both requires double maintenance and lots of duplicated code (#3535, #4034, we'd need to develop a moderation system for answers in addition to questions, etc etc). 

Let's make a few changes to the system, and consider merging the two types. We would have to:

1. make it possible to comment in response to another comment (perhaps using the comment.aid field to refer to a comment.cid it's in response to, instead of the answer.aid it's in response to)
2. convert all Answers into Comments, preserving timestamps, and changing all attached comments to the new Comment record, in a big database migration
3. display comments that respond to other comments [in the threaded way that Answers are currently shown](https://publiclab.org/questions/warren/02-21-2018/what-do-you-think-about-this-new-design-sketch-for-a-welcome-page)
4. make a "reply" button on comments that inserts the `cid` into the form so that the response goes to that comment, OR build in a commenting form to each comment... ?
5. migrate `answer_selection` (i.e. [likes on answers](https://github.com/publiclab/plots2/blob/master/app/models/answer.rb#L22)) to the differently-plumbed [comment likes](https://github.com/publiclab/plots2/blob/7740bee42c96cfdc32f188e4f69c98bd360a8064/app/models/comment.rb#L181-L207) (which is actually a nicer system with emoji and such)
6. figure out how the `question.answers` features will work, because answers can be "accepted" -- perhaps this also becomes part of the comment liking system, with a "checkbox" emoji? 
7. ensure statistics for Answers (http://publiclab.org/stats) are preserved somehow or merged into comments?

This is a big, complex project. But the outcome is that the codebase would be MUCH simpler and less redundant. There wouldn't be two different mixed types of responses to a post (comments + answers), but just one. We'd no longer have to have systems for turning comments into answers, and discussions wouldn't get fragmented across the two systems, as in https://publiclab.org/questions/warren/02-21-2018/what-do-you-think-about-this-new-design-sketch-for-a-welcome-page

This is a large project that will take careful work! Most of it may have to happen in the same pull request! 

We could conceivably make the comment-in-reply-to-comment system first, test it, then begin merging the old Answers into that system, to do this in 2 parts. But no matter what, it'll take some hard focused work to achieve and not lose any content.

### PR DESCRIPTION
**Pull request by ViditChitkara**
Fixes #4094

* [ ] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `rake test`
* [ ] code is in uniquely-named feature branch and has no merge conflicts 📁
* [ ] PR is descriptively titled 📑
* [ ] screenshots/GIFs are attached 📎 in case of UI updation
* [ ] ask `@publiclab/reviewers` for help, in a comment below

### CODE REVIEW

**Comment by SidharthBansal** [comment]
@ViditChitkara there are some conflicts, can you please rebase your changes?

**Comment by ViditChitkara** [comment]
@SidharthBansal sure, will do!!

**Comment by ViditChitkara** [comment]
@jywarren , the system is working fine but with each comment a "reply to this" comment box appears and according to me it is making our 'note' or 'question' page quite big. So, shall we hide the 'reply to comment' form in the beginning and just give a toggle link to open the reply form?

**Comment by jywarren** [comment]
Yeah, I think we can have just a toggle button to open the response form.
Can you screenshot once you change it?

Have you looked at how node.answers is counted because we track how many
responses there are to each question... Will that now just count comments?

Thanks!!!! Awesome!!!

**Comment by ViditChitkara** [comment]
Sure, will do once I change it. Answer count is indicated by
@node.answers.length now that answers will be deprecated, only comments
will count. Also the comment count will be a total of normal comments and
replied comments as well.
How does this sound??

Thanks

**Comment by jywarren** [comment]
Sounds great!

**Comment by ViditChitkara** [comment]
screenshot 2019-03-05 at 2 22 29 am
screenshot 2019-03-05 at 2 31 47 am
@jywarren here is how the toggle text will look like..this also shows how replied comments will be shown...what do you think?

PS: This is some test data on my local machine

**Comment by jywarren** [comment]
this looks great!

**Comment by ViditChitkara** [comment]
Cool, I'll be making a few changes and then I guess it will be good to go!!

**Comment by ViditChitkara** [comment]
@jywarren , I think a basic part of the issue is completed. Would you mind reviewing this PR and suggest any changes. Just wanted to know if I'm going in the right direction.

**Comment by ViditChitkara** [comment]
The parts which I think are completed are:-
1. Migration to convert answers to questions (I have still not deleted the answers which will be migrated. Maybe we can use them if needed in future).
2. Reply to comments feature. Added a reply_to field to comments and made 'aid' for all comments to be 0. Later on the aid column will be removed from comments.
3. Removed just the answer form and answers section from question/show page. This means that users will only be able to comment on a question.
@jywarren how does this sound?

**Comment by jywarren** [comment]
This is looking good - what parts do you think still need work?

**Comment by ViditChitkara** [comment]
I think we need to work on the following parts:-

1. migrate answer_selection (i.e. likes on answers) to the differently-plumbed comment likes (which is actually a nicer system with emoji and such). For this, will a like via "ThumbsUp" emoji do?

2. figure out how the question.answers features will work, because answers can be "accepted" -- perhaps this also becomes part of the comment liking system, with a "checkbox" emoji? 
We can surely add a "checkbox" emoji to comments which were earlier accepted answers, but since we will not be having answers in future, will this emoji make sense on comments?

3. ensure statistics for Answers (http://publiclab.org/stats) are preserved somehow or merged into comments?

@jywarren what do you think?
A lot of tests are failing due to usage of "reply_to" column in comments, working to fix that.

**Comment by jywarren** [comment]
OK. So, under the current state, we don't yet have to worry about stats for
answers because we aren't yet deleting any. Same with question.answers.

And people will still be able to post answers under the current state, or
have we turned that off?

And so the remaining issue would be to actually convert answers previously
posted into comments, and as part of that, make sure the answers stats and
questions.answers are both properly preserved?

Thank you, this is helpful!

**Comment by ViditChitkara** [comment]
@jywarren , we have turned off answer postings. Otherwise there will again be some answers which will have to be converted to comments. Other issues are complete with this pr I guess, so shall we test it somewhere?

**Comment by jywarren** [comment]
Yes, awesome, we can test on unstable - do you know how and do you need permissions to push to the unstable branch? 

Are we copying answer likes to be comment likes yet? I wouldn't want to lose existing information about those. 

Finally, I agree we should leave the answers table in place until we are 100% sure we have successfully copied all the information into the comments system with no losses. Then we can get rid of the table and model in follow-up!

Why don't you test it on unstable and see -- the other thing you can do is do `puts "Answers: " + Answer.count.to_s`  and "Comments: " + Comment.count.to_s` before and after the migration to have a record in the log of how many were converted successfully. How does that sound?

**Comment by ViditChitkara** [comment]
@jywarren , sure I'll test in on unstable. Yes, we are creating Comment Likes ("ThumbsUp emoji") for answer likes. However, nothing is deleted from the db. Even the comments which belong to answers will be available, since I'm just changing the reply_to field value in them, if anything wrong occurs we can easily get back to our older system.

**Comment by jywarren** [comment]
super!

**Comment by ViditChitkara** [comment]
I pushed this on unstable. It's working on unstable.publiclab.org . Kindly
have a look! I think it's ready to be merged now.

Thanks!

**Comment by jywarren** [comment]
Oh but it is failing for some reason? Can you take another look before we move forward? Thanks!!!

**Comment by ViditChitkara** [comment]
Sure, will have a look!

**Comment by ViditChitkara** [comment]
@jywarren the tests are passing now!! Have a look! Also added tests for replied comments.

File: public/cron_log.log [hunk 1]
```diff
@@ -0,0 +1,9 @@
+Thu Mar 14 18:41:00 UTC 2019
```
Comment (jywarren) [inline public/cron_log.log]: Oh oops, maybe we don't want to include this file?

**Comment by jywarren** [comment]
This is super! Sorry just one more file to remove, then this is ready to merge! 🎉

File: public/cron_log.log (hunk 1 above)
Comment (ViditChitkara) [reply public/cron_log.log]: Will do!

**Comment by ViditChitkara** [comment]
@jywarren done!

**Comment by jywarren** [comment]
Great! Let's check carefully on stable.publiclab.org once it builds there.

**Comment by ViditChitkara** [comment]
Have a look at the comments here. Some of them may have been answers which are converted to comments and corresponding comments are converted into replied comments!
https://stable.publiclab.org/questions/warren/05-20-2017/cheap-lightweight-gopro-alternatives-for-aerial-mapping
This is really exciting!

**Comment by ViditChitkara** [comment]
Should I start with 'removing answer related code' now as a follow-up of this?

**Comment by jywarren** [comment]
Sure, but wait, it seems like some answers don't exist on stable --
https://publiclab.org/questions/warren/05-20-2017/cheap-lightweight-gopro-alternatives-for-aerial-mapping#a849

was there some kind of error?

**Comment by ViditChitkara** [comment]
This is a bit odd!! At least the answer section should have been
removed.Also, I don't see the updated code over here. At least the comments
should have 'emoji like' feature which I think is not there. Will have a
look at it. Is there some way we could have a look at how the migrations
ran? Any sort of logs?
Thanks!!

**Comment by jywarren** [comment]
Yes, i think here?
https://jenkins.laboratoriopublico.org/job/Plots-Stable/lastBuild/console

**Comment by ViditChitkara** [comment]
Will have a look!

**Comment by ViditChitkara** [comment]
@Jeffrey Warren <jywarren@gmail.com> are we sure that the data on stable
was synced with production? Bcoz I do not see a 5 month old comment (which
is on production) over stable.
And the pattern suggests that last 3 answers (which are 1 year old are not
there). However, all older answers are converted to comments. Is there some
means by which we could sync the latest production data with stable and
then test?

**Comment by jywarren** [comment]
It's not fully synced, that's right! We could ask @icarito to do so, can
you contact him in the chatroom? Or, you could look on unstable, because
there should be the same "clone" of the database as stable, you know?

**Comment by ViditChitkara** [comment]
Actually I pushed this branch few days back to test on unstable so answers there would have been converted to comments. I guess we have got exactly the no. of comments here we want after conversion.
https://unstable.publiclab.org/questions/warren/05-20-2017/cheap-lightweight-gopro-alternatives-for-aerial-mapping#a849
However we can test it again on unstable. I'll ask @icarito about this!

**Comment by jywarren** [comment]
you could also look at https://stable.publiclab.org/comments to see what
the most recent comments are that were copied!

**Comment by ViditChitkara** [comment]
Will have a look!

**Comment by ViditChitkara** [comment]
I guess this is showing comments in descending order of creation timestamp. Since we have preserved answer's timestamp, we can't really figure out which one's were created today! Maybe we'll need to re-sync stable's data and then run the migration in this branch. @icarito any ideas?

>you could also look at https://stable.publiclab.org/comments to see what
the most recent comments are that were copied!

**Comment by jywarren** [comment]
OK, i think we are OK for this to go to production. I looked in the DB for the exact timestamps and this was one of the last comments in stable.publiclab.org before it was cloned: 

https://stable.publiclab.org/notes/gilbert/09-11-2017/hurricane-preparedness-dinner-discussion-sept-18-2017-6-pm-8-pm-cdt#c17361 -- it was on 12-09-2017, or Sept 12 2017. 

The missing comment by daisywatson here is from March 2018: https://publiclab.org/questions/warren/05-20-2017/cheap-lightweight-gopro-alternatives-for-aerial-mapping#a849

So it is not appearing because the db was copied before it was made. 

OK great!

**Comment by ViditChitkara** [comment]
Great!

**Comment by jywarren** [comment]
it's publishing now!

**Comment by ViditChitkara** [comment]
Fingers crossed!🤞

**Pull request merged**